src/core/insights/command-queue.ts

```diff
@@ -0,0 +1,137 @@
+  private queue: Command[] = [];
+  private debugMode: boolean = false;
+  private realImplementation: InsightsService | null = null;
+
+  constructor() {
+    // Create a proxy that will either queue commands or execute them directly
+    return new Proxy(this, {
+      get: (target: InsightsCommandQueue, prop: string) => {
+        // Return actual properties of the queue
+        if (
+          prop in target &&
+          typeof (target as InsightsCommandQueue)[prop] !== "function"
+        ) {
+          return (target as InsightsCommandQueue)[
+            prop as keyof InsightsCommandQueue
+          ];
+        }
+
+        // Return a function that either queues or executes the method call
+        return (...args: unknown[]) => {
+          if (!target.isInitialized || !target.realImplementation) {
+            // Queue the command for later execution
+            if (prop === "initInsights") {
+              // Special handling for initInsights - execute it right away
+              target.executeInitInsights(args[0] as InsightsConfig);
+              return;
+            }
+
+            // For debug logging
+            if (target.debugMode) {
+              logger.debug(`Queuing method call: ${String(prop)}`, args);
+            }
+
+            target.queue.push({
+              methodName: prop as keyof AnalyticsService,
+              args,
+            });
+            return;
+          }
+
+          // Execute the command immediately on the real implementation
+          if (
+            typeof (target.realImplementation as AnalyticsService)[prop] ===
+            "function"
+          ) {
+            return (target.realImplementation as AnalyticsService)[prop](
+              ...args,
+            );
+          }
+        };
+      },
+    });
+  }
+
+  // Special handling for init since it needs to happen right away
+  private executeInitInsights(config: InsightsConfig): void {
+    this.debugMode = !!config.debug;
+
+    if (this.debugMode) {
+      logger.debug("Initializing insights");
+    }
+
+    // Create and initialize the real implementation
+    this.realImplementation = new InsightsService();
+    this.realImplementation.initInsights(config);
+
+    // Mark as initialized and process the queue
+    this.isInitialized = true;
+    this.processQueue();
+  }
+
+  // Process all queued commands
+  private processQueue(): void {
+    if (this.debugMode) {
+      logger.debug(`Processing ${this.queue.length} queued commands`);
+    }
+
+    while (this.queue.length > 0) {
+      const cmd = this.queue.shift();
+      if (
+        cmd &&
+        this.realImplementation &&
+        typeof this.realImplementation[cmd.methodName] === "function"
+      ) {
+        try {
+          if (this.debugMode) {
+            logger.debug(
+              `Executing queued command: ${cmd.methodName}`,
+              cmd.args,
+            );
+          }
+          (
+            this.realImplementation[cmd.methodName] as (
```

Comment:
This is a bit of a tangle, I would break it up a bit as something like
```ts
const fn = this.realImplementation[cmd.methodName] as (
  ...args: unknown[]
) => unknown;

fn(...(cmd.args as unknown[]));
```

Reply:
Had to go with `fn.apply(this.realImplementation, cmd.args as unknown[])` to prevent those functions from falling over due to `this` no longer being the real implementation ¯\\\_(ツ)\_/¯